### ISSUE TITLE
Huge variability in the overwatch processing time

### ISSUE THREAD

**Issue by ruimaximo** [databrickslabs/overwatch#257]
Hi,

I am running overwatch 0.5.0.5 as a databricks job spinning its own cluster.
1. I frequently experience variations of overmatch processing time around 30%, even though there is no expected variation in the monitored clusters.
2. More rarely I experience increases of processing time 10 times folder.

![image](https://user-images.githubusercontent.com/32925235/137462057-5d6bae9e-673c-4b7b-8a6c-48ee43aef119.png)

Any explanation for the 1. and 2.? Any suggestions for mitigation measures?

Thank you

**Comment by mohanbaabu1996** [comment]
Hi @ruimaximo, overwatch is upgraded to 0.5.0.6, check the updates in the [docs](https://databrickslabs.github.io/overwatch/changelog/#0506). 
Shall we connect for a moment and i'll explain the steps to upgrade overwatch. I'm from IST time zone.

**Comment by ruimaximo** [comment]
It is solved by updating to version 0.5.0.6
Thank you guys for your support!

**Comment by GeekSheikh** [comment]
Hi @ruimaximo -- once per week there's an optimize and z-order that runs to maintain optimal data structures. 

The daily runtime deltas are also expected as there are two bottlenecks that are unavoidable. 

1. Spark Logs ingestion -- due to logs being stored in json.gz format. So there is a lot of skew and these must be trickle loaded. A lot of work has gone into optimizing this as much as possible but low CPU util is expected during the end of this step as the largest logs are loaded.
2. Cluster Event Logs Ingestion -- These are critical to the data set but the only place cluster event logs are made available are through a very inefficient API call. We must trickle-load these as well so as not to negatively affect other API requests on the platform. Low CPU utilization is expected during this step.

Both of the above bottlenecks are in bronze and are unavoidable at this time. There's an internal conversation going on to deliver these data sets in a more efficient fashion but we're still a ways out on this. We could discuss intelligent scaling and/or splitting bronze and silver/gold into separate jobs with separate configs if you like. If you'd like to explore this, please let us know some times and we'll schedule a meeting to walk through these.

**Comment by GeekSheikh** [comment]
@ruimaximo -- did you see my response above, does this help answer your question? Do you have more questions or should we close this ticket?

**Comment by ruimaximo** [comment]
Hi @GeekSheikh,

Thank you for your answer. I was waiting to have more data before answering to you.
Your answer makes sense but the bottleneck we observed is usually at the silver.  
We have an outlier of  > 11h per week that could be due to the table optimization. However, the bottleneck is at the silver (6h) and not at the bronze (42min) as you mentioned. Our total average processing time is around 3h per pipeline run. 

Note that this delta is not a major issue to us. I am trying to understand if the issue is somewhere else at Azure. It might be important to you if this means a potential underlying issue that you might not have experienced. I am available to assist you with more information if you find it relevant.

**Comment by GeekSheikh** [comment]
that is fairly unexpected, silver should not take that long. Could you please send me the log4j logs in an email or via slack for one of the longer runs? I'd like to take a look and see if I can figure out which step and / or why it's taking so stinking long. Please download all of the logs (there will be several files since it ran for several hours), zip them up, and shoot them over. Thanks

**Comment by GeekSheikh** [comment]
Thank you for sending the logs I have reviewed them and it's what I thought, it's the optimize and z-order that is taking a long time. I've created Issue #265 to allow this process to be pulled out of the standard pipeline. This will allow for the job that calls the optimize + zorder to have an autoscaling cluster that can grow larger so these processes can complete quicker and more efficiently.

Below is the timeline I found from the logs. I get why it seems that the silver ingestion is what's taking so long but the optimize/zorder occurs at the end of each pipeline. Customers can split out the pipelines and run them separately; thus each pipeline's tables (i.e. bronze/silver/gold) are always optimized upon completion of each pipeline.

14:03:33 -- Bronze Pipeline Begins -- no optimize necessary here since bronze has already been optimized in the past 7 days. Furthermore, there are very few z-orders in bronze thus this optimize is much quicker.
14:53:59 -- Silver Pipeline Begins
15:34:14 -- Silver Optimize & Zorder Begins -- Bronze and silver are already complete by this time
20:50:18 -- Gold begins
21:44:51 -- Gold Optimize & Zorder Begins -- Gold ingestion is complete by this point
The rest of the time in the pipeline is spent on the Gold Optimize & Zorder.
